tests/contracts/testMarket.nim

```diff
@@ -168,26 +173,66 @@ ethersuite "On-Chain Market":
     check receivedIds == @[request.id]
     await subscription.unsubscribe()
 
+  test "support request failed subscriptions":
+    await token.approve(storage.address, request.price)
+    await market.requestStorage(request)
+
+    var receivedIds: seq[RequestId]
+    proc onRequestFailed(id: RequestId) =
+      receivedIds.add(id)
+    let subscription = await market.subscribeRequestFailed(request.id, onRequestFailed)
+
+    for slotIndex in 0..<request.ask.slots:
+      await market.fillSlot(request.id, slotIndex.u256, proof)
+    for slotIndex in 0..request.ask.maxSlotLoss:
+      let slotId = request.slotId(slotIndex.u256)
+      while true:
+        try:
+          await waitUntilProofRequired(slotId)
+          let missingPeriod = periodicity.periodOf(await provider.currentTime())
+          await provider.advanceTime(periodicity.seconds)
+          await storage.markProofAsMissing(slotId, missingPeriod)
+        except JsonRpcProviderError as e:
```

Comment:
Thinking about this again, maybe it's more prudent to use `ProviderError` in case we decide to change which type of provider we use for tests in the future.

```suggestion
        except ProviderError as e:
```

Reply:
Agreed, I changed it.